BASE/src/ServerTelemetryChannel/Implementation/Transmitter.cs

```diff
@@ -169,10 +173,84 @@ internal virtual void Enqueue(Transmission transmission)
 
             if (!this.Storage.Enqueue(transmissionGetter))
             {
+                transmission.HasFlushTask = false;
                 TelemetryChannelEventSource.Log.TransmitterStorageSkipped(transmission.Id);
             }
         }
 
+        internal Task<bool> FlushAsync(Transmission transmission, CancellationToken cancellationToken)
+        {
+            TaskStatus taskStatus = TaskStatus.Canceled;
+            if (!cancellationToken.IsCancellationRequested)
+            {
+                transmission.HasFlushTask = true;
+                this.Enqueue(transmission);
+
+                try
+                {
+                    this.Storage.IncrementFlushAsyncCounter();
+                    taskStatus = this.MoveTransmissionsAndWaitForSender(transmission.FlushAsyncId, cancellationToken);
+                }
+                catch (Exception exp)
+                {
+                    taskStatus = TaskStatus.Faulted;
+                    TelemetryChannelEventSource.Log.TransmissionFlushAsyncWarning(exp.ToString());
+                }
+                finally
+                {
+                    this.Storage.DecrementFlushAsyncCounter();
+                }
+            }
+
+            Task<bool> flushTaskStatus = null;
+            if (taskStatus == TaskStatus.Canceled)
+            {
+                flushTaskStatus = TaskEx.FromCanceled<bool>(cancellationToken);
+            }
+            else if (taskStatus == TaskStatus.RanToCompletion && transmission.HasFlushTask)
+            {
+                flushTaskStatus = this.successTask;
+            }
+            else
+            {
+                flushTaskStatus = this.failedTask;
+            }
+
+            return flushTaskStatus;
+        }
+
+        internal Task<bool> MoveTransmissionsAndWaitForSender(CancellationToken cancellationToken)
+        {
+            if (cancellationToken.IsCancellationRequested)
+            {
+                return TaskEx.FromCanceled<bool>(cancellationToken);
+            }
+
+            var isStorageEnqueueSuccess = MoveTransmissions(this.Buffer.Dequeue, this.Storage.Enqueue, this.Buffer.Size);  
```

Comment:
Not sure where do we draw the line.. but this method is also a potentially long running (if high items and slow disk..), So should we pass cancellationtoken to it as well?

Reply:
We have to check the if cancellationtoken within the loop, this will impact performance as everytime we loop we need to check the status of cancellationtoken. If we are fine with this overhead I will add cancellationtoken check.

Reply:
As per our offline discussion, code has to check IsCancellationRequested in every logical step. Hence added cancelationtoken check.